### ISSUE TITLE
Default `buildPlugin()` and `buildPluginWithGradle()` to Java 11

### ISSUE THREAD

**Issue by NotMyFault** [jenkins-infra/pipeline-library#514]
Both calls still default to Java 8, when the minimum Java version to compile plugins against modern baselines require Java 11.

It would be nice to have Java 11 available as default, given that's the lowest supported Java version by now.

**Comment by lemeurherve** [comment]
I though about that but as @dduportal noted doing so would break all plugins compatible with Java 8 only (AFAIU).

**Comment by NotMyFault** [comment]
> I though about that but as @dduportal noted doing so would break all plugins compatible with Java 8 only (AFAIU).

Those would just need to specify JDK 8 then, no?

**Comment by lemeurherve** [comment]
> > I though about that but as @dduportal noted doing so would break all plugins compatible with Java 8 only (AFAIU).
> 
> Those would just need to specify JDK 8 then, no?

Yes, that all they need, but until they've done it (if ever), their plugin build will be broken.
We can't break an undetermined number of plugins like that (un)fortunately.

**Comment by dduportal** [comment]
> > > I though about that but as @dduportal noted doing so would break all plugins compatible with Java 8 only (AFAIU).
> > 
> > 
> > Those would just need to specify JDK 8 then, no?
> 
> Yes, that's all they need, but until they've done it (if ever), their plugin build will be broken. We can't break an undetermined number of plugins like that (un)fortunately.

+1 with what @lemeurherve said. However this task could be associated to a communication to all plugins developers + eventually a batch of PRs to change on plugins and maintainers could then choose to merge or not WDYT?

@NotMyFault worth an email on the dev mailing list to boostrap the subject?

**Comment by NotMyFault** [comment]
> eventually a batch of PRs to change on plugins and maintainers could then choose to merge or not

Yup, I remember we did similar stuff in the past to accomplish infrastructure changes, like switching to the reusable CD workflow.
Although, I'd imagine that not everyone who calls `buildPlugin()` only is stuck on Java 8 (thanks to major upgrade campaigns like [JENKINS-52012](https://issues.jenkins.io/browse/JENKINS-52012), [JENKINS-51805](https://issues.jenkins.io/browse/JENKINS-51805) and [JENKINS-52284](https://issues.jenkins.io/browse/JENKINS-52284)), the vast majority amount of people possible relies on the one-liner out of convenience, rather having a test matrix to run a subset of tasks on different JDKs.

> worth an email on the dev mailing list to boostrap the subject?

Yeah, communication is key. Albeit, I doubt there are many active and maintained plugins and components, which require compilation and execution on Java 8 only. 
This could also be seen as a chance for plugin authors to support more Java versions.

**Comment by lemeurherve** [comment]
Is there an easy way to identify plugins working only with jdk8?

**Comment by basil** [comment]
We are in the "phase change" portion of the migration now; i.e., the awkward period where both new and old worlds are  supported concurrently. At some point we will have to break with the past and drop support for Java 8 as a first-class citizen in the toolchain, including dropping support for Java 8 in things like the Jenkins test harness, acceptance test harness, Maven HPI plugin, plugin parent POM, etc. My preliminary thinking was that we should wait until at least the 2.375.1 LTS, given that [the developer documentation](https://www.jenkins.io/doc/developer/plugin-development/choosing-jenkins-baseline/) recommends that plugins support 1-3 LTS releases. The plugin parent POM currently supports cores all the way back to 2.249, but I think it is worth shorting our typical support cycle in this case. Unfortunately there is no foolproof way to detect if a plugin is compatible with Java 11 or not, but plugins with a parent POM earlier than 4.40 are unlikely to build and test cleanly against newer Jenkins cores on Java 11 without at least minor build changes.

**Comment by dduportal** [comment]
- Discussion in https://groups.google.com/g/jenkinsci-dev/c/pjfvsMw-EMM shows that there is not a clear consensus on this topic alas
- https://github.com/jenkins-infra/pipeline-library/pull/541 opened and closed for the reasons explained in the above email discussion
- documentation updates to recommend something else than the default `buildPlugin()` call in favor of an explicit list of configurations: 
  - https://github.com/jenkins-infra/pipeline-library/pull/542
  - https://github.com/jenkins-infra/jenkins.io/issues/5773

**Comment by jetersen** [comment]
Well, another way would be to do what plugin pom has done which is to pin the JDK based on Jenkins version?

So, if 2.361.1 use JDK 11.
If below assume JDK 8.

**Comment by jglick** [comment]
`archetypes` and docs tell you to use an explicit Java version in your `Jenkinsfile`, so beyond improving the library to warn you (or warn you more sternly?) when relying on the default, I think there is nothing to be done here.
